Content.Client/Chat/TypingIndicatorSystem.cs

```diff
@@ -0,0 +1,132 @@
+        /// </summary>
+        private bool _canTransmit = false;
+
+        public override void Initialize()
+        {
+            base.Initialize();
+            _canTransmit = true;
+            SubscribeNetworkEvent<RemoteClientTypingMessage>(HandleRemoteTyping);
+            SubscribeLocalEvent<PlayerAttachSysMessage>(HandlePlayerAttached);
+        }
+
+        public void HandleClientTyping()
+        {
+            if (!EnabledLocally || !_canTransmit) return;
+            RaiseNetworkEvent(new ClientTypingMessage(_playerManager?.LocalPlayer?.UserId, _attachedEntity?.Uid));
+            _canTransmit = false;
+            Timer.Spawn(1000 * (int) _cfg.GetCVar(CCVars.ChatTypingIndicatorPollRate), () =>
+            {
+                _canTransmit = true;
+            });
+        }
+
+        private void HandleRemoteTyping(RemoteClientTypingMessage ev)
+        {
+            var entity = EntityManager.GetEntity(ev.EnityId.GetValueOrDefault());
+            var comp = entity.EnsureComponent<TypingIndicatorComponent>();
+
+            //a remote client is typing. toggle their component so we can render the overlay locally.
+            comp.IsVisible = true;
+            Timer.Spawn(3000, () =>
+            {
+                comp.IsVisible = false;
+            });
+        }
+
+        private void HandlePlayerAttached(PlayerAttachSysMessage message)
+        {
+            _attachedEntity = message.AttachedEntity;
+        }
+
+        public override void FrameUpdate(float frameTime)
+        {
+            base.FrameUpdate(frameTime);
+            //this is ripped from debug health overlay. not good but it works.
+            if (_attachedEntity == null || _attachedEntity.Deleted)
+            {
+                return;
+            }
+
+            var viewBox = _eyeManager.GetWorldViewport().Enlarged(2.0f);
+
+            foreach (var typingIndicatorComp in EntityManager.EntityQuery<TypingIndicatorComponent>())
+            {
+                var entity = typingIndicatorComp.Owner;
+                if (_attachedEntity.Transform.MapID != entity.Transform.MapID ||
+                    !viewBox.Contains(entity.Transform.WorldPosition))
+                {
+                    if (_guis.TryGetValue(entity.Uid, out var oldGui))
+                    {
+                        _guis.Remove(entity.Uid);
+                        oldGui.Visible = false;
+                        oldGui.Dispose();
+                    }
+
+                    continue;
+                }
+
+                if (_guis.ContainsKey(entity.Uid))
+                {
+                    if (_guis.TryGetValue(entity.Uid, out var typGui))
+                    {
+                        typGui.Visible = typingIndicatorComp.IsVisible;
+                    }
+
+                    continue;
+                }
+
+                var gui = new TypingIndicatorGui(entity);
+                _guis.Add(entity.Uid, gui);
+            }
+        }
```

Comment:
Couldn't you just use a visualizer and attach it to the mob?

Reply:
so the chat bubble would be an entirely separate entity ? i thought of that at first but seemed wasteful

Reply:
nvm i guess entities can have multiple visualizers?